### PR TITLE
State events filtering database api

### PR DESCRIPTION
**Pull request by CromFr**
This PR adds a `gomatrixserverlib.Filter` parameter to functions handling the `syncapi_current_room_state` table. It does not implement any filtering logic inside the syncapi IncrementalSync/CompleteSync functions, just the APIs for future use.

Default filters are provided as placeholders in IncrementalSync/CompleteSync, so behaviour should be unchanged (except the default 20 event limit)

SQL table will be changed. You can upgrade an existing database using:
```sql
ALTER TABLE syncapi_current_room_state  ADD COLUMN IF NOT EXISTS sender text;
UPDATE syncapi_current_room_state SET sender=(event_json::json->>'sender');
ALTER TABLE syncapi_current_room_state ALTER COLUMN sender SET NOT NULL;
ALTER TABLE syncapi_current_room_state  ADD COLUMN IF NOT EXISTS contains_url bool;
UPDATE syncapi_current_room_state SET contains_url=(event_json::json->>'content')::json->>'url' IS NOT NULL;
ALTER TABLE syncapi_current_room_state ALTER COLUMN contains_url SET NOT NULL;
```

Note: This depends on #436 (and includes all its commits). I'm not sure if Github will remove the duplicated commits once #436 is merged.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/clientapi/routing/filter.go [hunk 1]
```diff
@@ -97,21 +89,21 @@ func PutFilter(
...
+	var filter gomatrixserverlib.Filter
 
 	if reqErr := httputil.UnmarshalJSONRequest(req, &filter); reqErr != nil {
 		return *reqErr
 	}
 
-	filterArray, err := json.Marshal(filter)
-	if err != nil {
+	// Validate generates a user-friendly error
+	if err = filter.Validate(); err != nil {
```
Comment (anoadragon453) [inline src/github.com/matrix-org/dendrite/clientapi/routing/filter.go]: Does this still check if the JSON is in the correct format, as the code did before?

File: src/github.com/matrix-org/dendrite/syncapi/routing/state.go [hunk 2]
```diff
@@ -44,7 +45,10 @@ func OnIncomingStateRequest(req *http.Request, db *storage.SyncServerDatabase, r
 	// TODO(#287): Auth request and handle the case where the user has left (where
 	// we should return the state at the poin they left)
 
-	stateEvents, err := db.GetStateEventsForRoom(req.Context(), roomID)
+	stateFilterPart := gomatrix.DefaultFilterPart()
```
Comment (anoadragon453) [inline src/github.com/matrix-org/dendrite/syncapi/routing/state.go]: A few `gomatrix`'s still left in here.

File: src/github.com/matrix-org/dendrite/syncapi/storage/current_room_state_table.go [hunk 3]
```diff
@@ -166,9 +178,17 @@ func (s *currentRoomStateStatements) selectRoomIDsWithMembership(
 // CurrentState returns all the current state events for the given room.
 func (s *currentRoomStateStatements) selectCurrentState(
 	ctx context.Context, txn *sql.Tx, roomID string,
+	stateFilterPart *gomatrix.FilterPart,
```
Comment (anoadragon453) [inline src/github.com/matrix-org/dendrite/syncapi/storage/current_room_state_table.go]: and here

File: src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go [hunk 4]
```diff
@@ -177,10 +178,10 @@ func (d *SyncServerDatabase) GetStateEvent(
 // Returns an empty slice if no state events could be found for this room.
 // Returns an error if there was an issue with the retrieval.
 func (d *SyncServerDatabase) GetStateEventsForRoom(
-	ctx context.Context, roomID string,
+	ctx context.Context, roomID string, stateFilterPart *gomatrix.FilterPart,
```
Comment (anoadragon453) [inline src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go]: and here

File: src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go [hunk 5]
```diff
@@ -233,11 +234,13 @@ func (d *SyncServerDatabase) IncrementalSync(
 	var succeeded bool
 	defer common.EndTransaction(txn, &succeeded)
 
+	stateFilterPart := gomatrix.DefaultFilterPart() // TODO: use filter provided in request
```
Comment (anoadragon453) [inline src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go]: and here

File: src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go [hunk 6]
```diff
@@ -286,11 +289,13 @@ func (d *SyncServerDatabase) CompleteSync(
 		return nil, err
 	}
 
+	stateFilterPart := gomatrix.DefaultFilterPart() // TODO: use filter provided in request
```
Comment (anoadragon453) [inline src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go]: and here

File: src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go [hunk 7]
```diff
@@ -547,6 +552,7 @@ func (d *SyncServerDatabase) fetchMissingStateEvents(
 func (d *SyncServerDatabase) getStateDeltas(
 	ctx context.Context, device *authtypes.Device, txn *sql.Tx,
 	fromPos, toPos types.StreamPosition, userID string,
+	stateFilterPart *gomatrix.FilterPart,
```
Comment (anoadragon453) [inline src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go]: and here

File: vendor/src/github.com/matrix-org/gomatrix/filter.go [hunk 8]
```diff
@@ -14,30 +14,77 @@
...
+	NotSenders  []string `json:"not_senders,omitempty"`
+	NotTypes    []string `json:"not_types,omitempty"`
+	Senders     []string `json:"senders,omitempty"`
+	Types       []string `json:"types,omitempty"`
+	ContainsURL *bool    `json:"contains_url,omitempty"`
+}
+
+// Validate checks if the filter contains valid property values
+func (filter *Filter) Validate() error {
+	if filter.EventFormat != "client" && filter.EventFormat != "federation" {
```
Comment (anoadragon453) [inline vendor/src/github.com/matrix-org/gomatrix/filter.go]: Is this still relevant to the client-side `gomatrix` lib?

File: src/github.com/matrix-org/dendrite/clientapi/routing/filter.go (hunk 1 above)
Comment (CromFr) [reply src/github.com/matrix-org/dendrite/clientapi/routing/filter.go]: What do you mean by "if the JSON is in the correct format" ?

`UnmarshalJSONRequest` already checks if the JSON is valid and that types from the Filter struct match types from the json data.
`Validate()` only checks if values inside the Filter struct match those allowed by the documentation  (ie the `event_format` field)

File: vendor/src/github.com/matrix-org/gomatrix/filter.go (hunk 8 above)
Comment (CromFr) [reply vendor/src/github.com/matrix-org/gomatrix/filter.go]: IMHO it's still interesting to have some kind of client-side validation before sending the query.
Though it's not very useful since you must rely on server response anyway.

File: src/github.com/matrix-org/dendrite/clientapi/routing/filter.go (hunk 1 above)
Comment (anoadragon453) [reply src/github.com/matrix-org/dendrite/clientapi/routing/filter.go]: Ah, right. I was just looking at the `Marshal` being taken off and thinking that was the only line checking the JSON correctness.

File: vendor/src/github.com/matrix-org/gomatrix/filter.go (hunk 8 above)
Comment (anoadragon453) [reply vendor/src/github.com/matrix-org/gomatrix/filter.go]: Ah, well best a chat left to the gomatrix repo anyways.

**Comment by anoadragon453** [comment]
LGTM.

**Comment by Cnly** [comment]
Since this PR has got an approval but stuck for a really long time, I'd like to help merge master onto it and see what we'll get.
